Genesis of RAML based Rest API documentation.

aquamatthias opened the pull request:
- [x] Add intro raml documentation 
- [x] Add a link to the UI
- [x] Question: there is a raml2html tool which we could use to create one static html file for the docs folder. Do you think this is better?
- [x] Remove the "try it out" button from the static documentation

meichstedt commented on docs/docs/rest-api/public/api/api.raml:
```diff
@@ -0,0 +1,29 @@
+#%RAML 0.8
+
+title: Marathon REST API
+version: 2.0
```
This value is not rendered in the output html. The entries are rather self-explanatory since they include the `/v2/` path element, but I think this should be used to render `Marathon REST API 2.0`

aquamatthias replied:
Not exactly sure what to do here. Add 2.0 to the title?

meichstedt commented on docs/docs/rest-api/public/api/api.raml:
```diff
@@ -0,0 +1,29 @@
+
+documentation:
+ - title: Using the Marathon REST API
+   content: Make me real!
```
I need to click on the title in order to open a layover that renders the content. Feels like there should be text rendered directly underneath `Documentation` – expanding the `Using the Marathon REST API` makes no sense, and I don't think we need too much explanation for the documentation header.

meichstedt commented on docs/docs/rest-api/public/api/traits.raml:
```diff
@@ -0,0 +1,15 @@
+- secured:
+    responses:
+      401:
+        description: Authentication is enabled and you did not provide enough or wrong information to authenticate that request.
```
I'd place * Unauthorized* in front, so one does not have to read the whole sentence ;)

meichstedt commented on docs/docs/rest-api/public/api/traits.raml:
```diff
@@ -0,0 +1,15 @@
+            example: |
+              { "message": "description for authentication failed" }
+      403:
+        description: Authorization is granted but the identity provided does not have sufficient access rights to do that action.
```
As above: **Forbidden** in front of the remaining description

meichstedt commented on docs/docs/rest-api/public/api/traits.raml:
```diff
@@ -0,0 +1,15 @@
+        body:
+          application/json:
+            example: |
+              { "message": "description for autorization failed" }
```
Are the messages placeholders? Should they contain the actual message with which we respond, or do we respond with different messages? If we do not document the exact message (I guess we won't), I'd put the message in `<>` braces.

aquamatthias replied:
The problem here is, that this message is generated from a plugin. But I agree if this is an example, we should make it one.

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+# Get a list of applications
+#
+get:
+  description: Get the list of running applications. Several filters can be applied to filter the result list.
```
... via the following query parameters.

meichstedt replied:
Should we mention that the call is/might get expensive? Is there a raml-way of adding a note/mark other than mis-using traits for that?

aquamatthias replied:
The description can be defined as markdown. So you can highlight anything you want in the description. There is no specific note element.

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+  is: [ secured ]
+  queryParameters:
+    cmd:
+      description: the command of this application
```
"Filter the result to only return apps whose `cmd` field contains the given `cmd`, Default: `""`"

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+      description: the command of this application
+      example: java
+    id:
+      description: the id or part of the id of the application
```
"Filter the result to only return apps whose `id` is or contains the given value"

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+
+        * <code>Set based Selector Query</code> Matches existence of labels and the (non) existence of the value in a given set
+
+      example: environment==production, tier!=frontend\ tier, deployed in (us, eu), deployed notin (aa, bb)
```
An example for an existence based selector is missing; only equality and set based are given. As far as I understand this, setting label to `environment` would match apps that have an `environment` set – no matter to which value.

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+#
+get:
+  description: Get the list of running applications. Several filters can be applied to filter the result list.
+  is: [ secured ]
```
As far as I can see, this leads to the output having a `Security Schemes` section with one `Anonymous` entry – this might be a little irritating, as it looks like _the anonymous security scheme is applied_ …

aquamatthias replied:
The security schema is a different thing - with that you can define how to access the API (e.g. basic auth). The secured trait is attached to all endpoints, that would need that behavior.

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+        You can specify this parameter multiple times with different values.
+
+        - <code>apps.tasks</code> Apps tasks are not embedded in the response
+        by default. This currently implies apps.deployments currently but that
```
drop one of the two _currently_'s – I'd hunt down the second one.

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+      example: environment==production, tier!=frontend\ tier, deployed in (us, eu), deployed notin (aa, bb)
+    embed:
+      description:
+        Embeds nested resources that match the supplied path.
```
I figure this is copied from the former docs, but in general, I think it'd be easier to read if all flags followed an imperative (_eek_) style along with italic annotations, like 
`"apps.counts"`: embed task counts (tasksStaged, tasksRunning, tasksHealthy, tasksUnhealthy). _Note: these are currently embedded by default but this will change in a future release._

aquamatthias replied:
Like the idea. Will rewrite that part.

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+# Create application
+#
+post:
+  description: Create and start a new application.
```
In the old API docs, we have documented most of the body-JSON-properties as well – this is and supposedly cannot be integrated into the raml docs because raml does not support documentation ob body fields. I don't think this is a problem, it's even better than having properties described in the schema _and_ in the docs. We would need to make sure that all properties are correctly described in the JSON schema, though.

aquamatthias replied:
I have started to integrate this documentation into the json schema. It would be the natural place for raml documentation. Unfortunately there is no nice visualization. So my suggestion would be to add the model description to the introduction part of the documentation.

meichstedt replied:
Ok, I think that would be fine as well.

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+      description: The application definition provided in the body is not valid.
+      body:
+        application/json:
+          example: |
```
Maybe the example should be short and only show a basic example – and the schema open by default? If the example is long, the schema-button is a long way down the page.

aquamatthias replied:
I changed that example back and forth: 
- I think most of the people will not try to understand json schema
- I think an example is better than the schema
- I think a short example is easy to understand, but is always only half of the truth.

I decided to use a real example here.

meichstedt replied:
I agree, a real world example probably explains best. (Additionally, if we get the fancy strict mode implemented, then the result of trying out a JSON snippet will speak for itself).

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+  # Delete a specific app
+  #
+  delete:
+    description: Destroy an application. All data about that application will be deleted.
```
Shouldn't we document/explain what exactly happens, and that there might still be running tasks when the response is returned?

aquamatthias replied:
Added a sentence with respect to deployments.

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+    is: [ secured ]
+    responses:
+      204:
+        description: The response will contain an empty body.
```
Actually that's a 2012 and it will contain a deployment result.

aquamatthias replied:
Good catch.

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+  ###
+  # Delete a specific app
+  #
+  delete:
```
The `force` parameter is undocumented

aquamatthias replied:
Made a deployable trait, which documents the force flag and added that to all related endpoints.

meichstedt replied:
:+1:

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+                { "message": "App '/not_existent' does not exist" }
+
+    ###
+    # Create application
```
s/Create/Delete

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+               "deployments":[{"id":"97c136bf-5a28-4821-9d94-480d9fbb01c8"}]
+              }
+
+  /restart:
```
undocumented: `force` query param

aquamatthias replied:
Add deployable trait.

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+          description: If `scale=true` is specified, then the application is scaled down by the number of killed tasks.
+          default: false
+      responses:
+        200:
```
Actually, the response is different depending on the scale param:
- if `scale == true`, the result will be a deployment info
- otherwise, the result will be a list of tasks

aquamatthias replied:
True. Added that.

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+            default: false
+        responses:
+          200:
+            description: The tasks that has been killed.
```
- if `scale == true`, a deployment result
- otherwise _the_ task that has been killed (in a field called `task`)

aquamatthias replied:
Added that.

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+
+    /{task_id}:
+      ###
+      # Create application
```
s/Create application/Delete one task

aquamatthias replied:
Done.

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,451 @@
+                  {"message":"Task 'not-existing' does not exist"}
+
+  ###
+  # Create application
```
s/Create/List versions of an

I get the feeling that there is the same amount of misleading comments and of correct ones :see_no_evil: 
Do we need really need them?

aquamatthias replied:
While writing this, they help me find the single endpoints more easily. I corrected all - hope this helps also in the future.

meichstedt commented on docs/docs/rest-api/public/api/v2/groups.raml:
```diff
@@ -0,0 +1,376 @@
+
+    * Changes to application parameters will result in a restart of this application.
+
+    * A new application added to the group is started.
```
s/is/will be

meichstedt commented on docs/docs/rest-api/public/api/v2/groups.raml:
```diff
@@ -0,0 +1,376 @@
+
+    * A new application added to the group is started.
+
+    * An existing application removed from the group gets stopped.
```
s/gets/will be

meichstedt commented on docs/docs/rest-api/public/api/v2/groups.raml:
```diff
@@ -0,0 +1,376 @@
+    If there are no changes to the application definition, no restart is triggered.
+    During restart marathon keeps track, that the configured amount of minimal running instances are _always_ available.
+
+    This method alows 2 special modes for the update operation>
```
s/alows/allows

meichstedt replied:
Add: When one of `version` or `scaleBy` are set, nothing else than a version change or transitive instance count scaling will be applied. If both `version` and `scaleBy` are set, only a version rollback will be performed – the scaleBy value will not be applied.

- There should be tests for this implementation detail
- It might be better to throw validation errors than to silently drop body parts by performing `versionChange orElse scaleChange getOrElse update.apply`.

aquamatthias replied:
Added. I agree with that checks - should be part of validation though.

meichstedt commented on docs/docs/rest-api/public/api/v2/groups.raml:
```diff
@@ -0,0 +1,376 @@
+
+    * Provide only the `version` field in the group definition. This will rollback the group to that given version
+
+    * Provide only the `scaleBy` field will scale all transitive applications by the given factor.
```
s/applications/applications' instance counts

Also: an explanation how rounding will be applied might be helpful. Imagine I have a group with 3 apps:
- app-1, instance count: 100
- app-2, instance count: 100
- app-3, instance count: 1

people will ask (themselves or us): what happens to app-3 when I scaleBy 0.1?

meichstedt commented on docs/docs/rest-api/public/api/v2/groups.raml:
```diff
@@ -0,0 +1,376 @@
+
+    * Provide only the `scaleBy` field will scale all transitive applications by the given factor.
+
+    A deployment can run forever. This is the case, when the new application has a problem and does not become healthy.
```
This paragraph is rendered relative to the above bullet point which, I guess, is unintended. You need to insert another blank line above to fix this …

meichstedt replied:
This can also have other reasons, e.g. lack of suitable resource offers. In that case, the deployment will stay in state `Waiting`, and there's no problem in terms of an error in the AppDefinition.

In general, there's a lot of prose written here. We should consider moving prose to the general documentation and only documenting the rough edges here, while maintaining links from either part to the other. Of course, we might decide to put everything in here, but I think that would be beyond the scope of a REST API doc.

aquamatthias replied:
1) Thanks for looking into the rendered result. The markdown parser is rather clumsy...
2) Part of the problem is also: that the endpoint is able to do __a lot__

Generally I agree: we want to have short/sharp reference docs and longer prose as introduction.

meichstedt commented on docs/docs/rest-api/public/api/v2/groups.raml:
```diff
@@ -0,0 +1,376 @@
+
+
+    If there is an upgrade process already in progress, a new update will be rejected unless the force flag is set.
+    With the force flag given, a running upgrade is terminated and a new one is started.
```
`force` should be documented via `queryParameters:`

meichstedt replied:
We should describe somewhere, what happens with tasks created during a later on cancelled deployment :speak_no_evil:

aquamatthias replied:
added deployable

meichstedt commented on docs/docs/rest-api/public/api/v2/groups.raml:
```diff
@@ -0,0 +1,376 @@
+    If there is an upgrade process already in progress, a new update will be rejected unless the force flag is set.
+    With the force flag given, a running upgrade is terminated and a new one is started.
+
+    Since the deployment of the group can take a considerable amount of time, this endpoint returns immediatly with a version.
```
s/immediatly/immediately

aquamatthias replied:
Done.

meichstedt commented on docs/docs/rest-api/public/api/v2/groups.raml:
```diff
@@ -0,0 +1,376 @@
+      description:
+        A deployment is started which has a unique deployment identifier.
+        The related deployment can be fetched from the /v2/deployments endpoint.
+        If the deployement is gone from the list of deployments, than this means it is finished.
```
s/deployement/deployment

aquamatthias replied:
removed.

meichstedt commented on docs/docs/rest-api/public/api/v2/groups.raml:
```diff
@@ -0,0 +1,376 @@
+    With the force flag given, a running upgrade is terminated and a new one is started.
+
+    Since the deployment of the group can take a considerable amount of time, this endpoint returns immediatly with a version.
+    The failure or success of the action is signalled via event. There is a group_change_success and group_change_failed with
```
s/event/events

meichstedt commented on docs/docs/rest-api/public/api/v2/groups.raml:
```diff
@@ -0,0 +1,376 @@
+        application/json:
+          example: |
+            {"message":"Invalid JSON","details":[{"path":"/id","errors":["error.expected.jsstring"]}]}
+    409:
```
The 409 does not belong here but only to the POST section. A PUT will always update the group as long as there is no validation error.

aquamatthias replied:
Done.

meichstedt commented on docs/docs/rest-api/public/api/v2/groups.raml:
```diff
@@ -0,0 +1,376 @@
+            }
+
+
+/{group_id}:
```
docs for /group/{group_id} are the same as for /group – I'd extract those into a separate raml file and include that in both cases. Should we point out that you can shoot against nested groups?

aquamatthias replied:
Good idea. Removed everything under {group_id} and imported the same file twice.

meichstedt commented on docs/docs/rest-api/public/api/v2/groups.raml:
```diff
@@ -0,0 +1,376 @@
+###
+# Create a new application group
+#
+post:
```
The underlying implementation has a `force` flag that is undocumented but will allow changing the group even if there's a running deployment.

aquamatthias replied:
forceable -> deployable

meichstedt replied:
That's a better name indeed.

meichstedt commented on docs/docs/rest-api/public/api/v2/tasks.raml:
```diff
@@ -0,0 +1,43 @@
+  queryParameters:
+    status:
+      description: Filter the list of tasks by status
+      enum: [ TASK_STAGING, TASK_STARTING, TASK_RUNNING, TASK_FINISHED, TASK_FAILED, TASK_KILLED, TASK_LOST, TASK_ERROR ]
```
Not quite. According to the [old docs](https://mesosphere.github.io/marathon/docs/rest-api.html#get-v2-tasks) and [the implementation](https://github.com/mesosphere/marathon/blob/d098e5f73905d91e9511dbbdf970472656d4649b/src/main/scala/mesosphere/marathon/api/v2/TasksResource.scala#L135-L139), possible values are `[running,staging]`

aquamatthias replied:
In theory we could report tasks in other states. From an API perspective that could make sense. In order to document the current status I agree.

meichstedt replied:
1) the implementation will match the given string like this
```
    case "running" => Some(TaskState.TASK_RUNNING)
    case "staging" => Some(TaskState.TASK_STAGING)
    case _         => None
```
so passing `TASK_STAGING ` will **not** work but return all tasks disregarding the status. In that sense, the documentation is still wrong.

2) Yes, the base for the result of this endpoint is the TaskTracker list, so integrating any other task states would be quite easy.

Please fix 1)

meichstedt commented on docs/docs/rest-api/public/api/v2/tasks.raml:
```diff
@@ -0,0 +1,43 @@
+      repeat: true
+  responses:
+    200:
+      description: The list of running tasks.
```
The list of all tasks disregarding their status, or a list of tasks matching the specified `status` filter.

meichstedt commented on docs/docs/rest-api/public/api/v2/tasks.raml:
```diff
@@ -0,0 +1,43 @@
+          { "ids": [ "task1", "task2" ] }
+    responses:
+      200:
+        description: No content is send with this reponse.
```
Should we document this: if `scale == true`, a `DeploymentPlan` will be created for each app corresponding to one or more of the given tasks. These are not returned, so one cannot track the deployments in order to see when the killing of multiple tasks with scaling eventually finished.

btw: I am not convinced that this functionality makes sense and it used (killing tasks of multiple apps at once while scaling their instance count to whatever remains). Maybe we should deprecate that and see if anyone complains.

aquamatthias replied:
Added documentation for the 2 options.

meichstedt commented on docs/docs/rest-api/public/api/v2/events.raml:
```diff
@@ -0,0 +1,33 @@
+get:
+  description:
+    Attach to the marathon event stream.
```
The `Try it` section for events is not working

aquamatthias replied:
This is a question of your expectation: 
- if you add the Accept Header, the Try it out function connects to the event stream and will not return -> you see nothing
- if you omit the Accept Header, you get a 405 response.

That means this functionality can not be used in the console. 
I will add a sentence to the docs.

meichstedt commented on docs/docs/rest-api/public/api/v2/queue.raml:
```diff
@@ -0,0 +1,34 @@
+
+  delete:
+    description:
+      If an application fails to often in a specified amount of time (according to the application definition),
```
s/to/too

meichstedt commented on docs/docs/rest-api/public/api/v2/queue.raml:
```diff
@@ -0,0 +1,34 @@
+      The effect is, that the tasks of this application will be launched immediately.
+    responses:
+      200:
+        description: the delay is resetted and no body is returned
```
s/resetted/reset

meichstedt commented:
Phew, that was a lot to comment, but not too many critical things. In general, I like the structure and maintainability, along with the option to `try it out`.

- several sections contain misleading or wrong information, at least some of those were ported from the old doc
- In various locations, there is more info than just API documentation. That's not bad at all, maybe even better than having these separately in the _general_ documentation. However, having a lot of prose in here creates the need to read a lot, and prevents from quickly scanning through available params. Therefore, I'd rather move too detailed explanations into prose based docs and let the API docs only outline the API. If we do that, then:
  - there ought to be a lot of very specific cross linkages, so people can find the API doc to a given feature, or vice versa
  - we need to provide generated html with anchors for specific endpoints so we can directly link them from the general documentation.

Using raml2html for generating static docs without the `Try it` would make sense. Those static pages should reside on the web site relative to the version for which it was generated. I'm thinking about `/0.13/v2/api` vs `/0.14/v2/api` vs `/0.14/v3/api`, so that there's a specific documentation for each Marathon version and API version.

Please fix the commented issues and let's discuss how we can incorporate this!

aquamatthias commented:
Cool - had the same idea with versioning.

aquamatthias commented:
@meichstedt Thanks for all the good hints! I hope I addressed them all.
I still feel there is a way for really good documentation, but at least it feels like a good start.

Changes:
- Incorporated (hopefully) all suggestions
- biggest change probably is the usage of the deployable trait (made a lot things easier - thanks for the suggestion)
- removed api-console from the docs folder and use the static generated file (not versioned for now)
- index: moved current rest-api to rest-api (deprecated) and add the new rest-api 
- deployed version 0.1.0 of api-console and use this as runtime dependency for /help

meichstedt commented on docs/docs/rest-api/public/api/v2/apps.raml:
```diff
@@ -0,0 +1,416 @@
+
+        - <code>apps.tasks</code> embed all tasks of each application<br/>
+        Note&#58; if this embed is definded, it automatically sets <code>apps.deployments</code> but this will change in a future release.
+        Please define all embeds explicit.
```
Good idea to add that hint. However, please `s/explicit/explicitly`, as this is an adverb to `define`. This does also hold for the 2 occurrences below.

meichstedt commented:
Hey @aquamatthias looks fine – just 2 little issues left:
- wrong grammar in `apps.raml`
- wrong documentation of the `status` queryParam in `tasks.raml`

I made some more comments but as far as I can see these are the only 2 with need for action.

The latter one prevents me from merging, please fix that.

aquamatthias commented:
@meichstedt addressed your comments. Thanks. If you agree, I will squash the commits.

meichstedt commented:
@aquamatthias perfect – please squash and wait for the sound of the foghorn :ship:
